docs/rendering.rst

```diff
@@ -0,0 +1,783 @@
+is not defined, and outputs an empty string, leaving ``<p></p>`` in the
+preview output.
+
+Other macros don't have specific code to detect preview mode, and have
+kumascript rendering errors.
+
+Some macros, like ``{{HTMLElement}}``, can work as expected in preview.
+
+Raw content
+===========
+A ``?raw`` parameter can be added to the end of a document to request the
+source for a revision.
+
+source
+   *cleaned content*, with filters
+output
+   The page with a ``?raw`` query parameter
+database
+   *not stored*
+code
+   *not available*
+
+For the simple document, this is the *raw content*:
+
+.. code-block:: html
+
+   <p>{{ CSSRef }}</p>
+
+   <p>I am a <strong>simple document</strong> with a CSS sidebar.</p>
+
+   <p style="color: red;">I am red.</p>
+
+   <h2 id="Some_Links">Some Links</h2>
+
+   <ul>
+    <li><a href="/en-US/docs/Web/HTML">The HTML Reference</a></li>
+    <li>{{HTMLElement('div')}}</li>
+    <li><a href="/en-US/docs/NewDocument">A new document</a></li>
+   </ul>
+
+   <div class="button"></div>
+
+   &lt;script&gt;
+     alert('How about this?');
+   &lt;/script&gt;
+
+The *cleaned content* is parsed for filtering . The headers get IDs, based on
+the content, if they did not have them before.  For example,
+``id="Some_Links"`` is added to the ``<h2>``.
+
+A simple filter is applied that strips any attributes that start with
+``on``, such as the scripting attempt ``onclick``. However, none of these
+should remain in the *cleaned content*.
+
+Live sample
+============
+`Live samples`_ are stored in document content. The content is then processed
+to extract the CSS, JS, and HTML, and reformat them as a stand-alone HTML
+document suitable for displaying in an ``<iframe>``.
+
+source
+   A section extracted from *rendered content*, with further processing
+output
+   Live sample documents on a separate domain, such as https://mdn.mozillademos.org
+database
+   Not stored in the database, but cached
+code
+   ``Document.extract_code_sample(section_id)``
+
+`Live samples`_ are long, so the simple document does not include one.
+
+.. _`Live samples`: https://developer.mozilla.org/en-US/docs/MDN/Contribute/Structures/Live_samples
+
+Future Changes
+==============
+Rendering evolved over years, and this document describes how it works, rather
+than how it was designed. There are some potential changes that would simplify
+rendering:
+
+* Sidebar macros are heavy users of API data and require post-processing of the
+  content. Sidebar generation could be moved into Kuma instead of being
+  specified by a macro.
+* The *diff format* could be replaced by the *cleaned content* format, which
+  would be stored for each revision rather than just for the most recent
+  document.
+* Content from editing could be normalized and filtered before storing as the
+  *revision content*. This may unify the *re-edit format*, *diff format*, and
+  *cleaned content*
+* Add IDs immediately to the *revision content*, rather than wait for the
```

Comment:
Nit: you switch out of subjunctive mode here.  Maybe start this bullet and the next with "We could add"